### PR TITLE
prompt add a shortcut `c-s` to pick a word under cursor. command `search` make histories as completions.

### PR DESCRIPTION
**Pull request by cossonleo**
Like title
ui prompt add a shortcut to pick a word.
Command `search` make histories as completions
The `global search` currently has no completions because it does't have histories register.

### CODE REVIEW

**Comment by cossonleo** [comment]
> Oh, I don't think I have seen this in other text editors but seemed like an interesting idea.
> 
> But wouldn't a history completion be more useful? Like fish.

In vim/neovim, we can use <c-r><c-w> to fast pick the word under cursor. But I think using completions is more ergonomics.
There is shortcut for picking history. I don't know it is unnecessary for add history to completions.

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -1158,6 +1160,47 @@ fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Rege
     };
 }
 
+fn search_completions(cx: &mut Context, reg: Option<char>) -> Vec<String> {
+    let count = cx.count();
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+
+    let selection = doc.selection(view.id).clone().transform(|range| {
+        textobject::textobject_word(text, range, textobject::TextObject::Inside, count)
```
[the remark this answers is not in the document]
Comment (pickfire) [reply helix-term/src/commands.rs]: Yup that is good for now, later I can make it as ctrl-r.

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -1169,6 +1171,26 @@ fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Rege
...
+                    } else {
+                        exist_elems.insert(item);
+                        true
+                    }
+                })
+                .take(10)
+                .cloned()
+                .collect()
+        })
+}
```
Comment (pickfire) [inline helix-term/src/commands.rs:1192]: I don't get why you need to `take(10)` or iterate through it when `cx.editor.registers.get(reg)` only return an option. It's just one item at most.

**Comment by pickfire** [comment]
Can you please update the book for this key as well? Also, can also add a comment there that this will probably be changed to `ctrl-r w` later (that will probably be done together with an infobox).

File: helix-term/src/commands.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cossonleo) [reply helix-term/src/commands.rs]: Already add tit o book

**Comment by cossonleo** [comment]
Already add it to book

**Comment by archseer** [comment]
Sorry for the delay, I will test the PR later today!

File: book/src/keymap.md [hunk 3]
```diff
@@ -258,3 +258,25 @@ Keys to use within picker. Remapping currently not supported.
...
+| `Alt-b`, `Alt-Left`     | Backward a word                                                 |
+| `Ctrl-b`, `Left`        | Backward a char                                                 |
+| `Alt-f`, `Alt-Right`    | Forward a word                                                  |
+| `Ctrl-f`, `Right`       | Forward a char                                                  |
+| `Ctrl-e`, `End`         | move prompt end                                                 |
+| `Ctrl-a`, `Home`        | move prompt start                                               |
+| `Ctrl-w`                | delete previous word                                            |
+| `Ctrl-k`                | delete to end of line                                           |
+| `backspace`             | delete previous char                                            |
+| `Ctrl-s`                | insert a word under doc cursor, maybe changed to Ctrl-r Ctrl-w  |
```
Comment (pickfire) [inline book/src/keymap.md:276]:
```suggestion
| `Ctrl-s`                | insert a word under doc cursor, may be changed to Ctrl-r Ctrl-w later  |
```

**Review by pickfire — APPROVED** [review]
Nice, looks good to me. Only left some minor text fix on the doc.

File: helix-term/src/commands.rs [hunk 4]
```diff
@@ -1179,11 +1202,19 @@ fn search(cx: &mut Context) {
...
 
     let prompt = ui::regex_prompt(
         cx,
         "search:".into(),
         Some(reg),
+        move |input: &str| {
+            completions
+                .iter()
+                .filter(|comp| comp.starts_with(input))
+                .map(|comp| (0.., std::borrow::Cow::Owned(comp.clone())))
```
Comment (archseer) [inline helix-term/src/commands.rs:1215]: Could we use `Cow::Borrowed` here?

File: helix-term/src/commands.rs [hunk 5]
```diff
@@ -1169,6 +1171,27 @@ fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Rege
...
+                .iter()
+                .rev()
+                .filter(|&item| {
+                    if exist_elems.contains(item) {
+                        false
+                    } else {
+                        exist_elems.insert(item);
+                        true
+                    }
+                })
```
Comment (archseer) [inline helix-term/src/commands.rs:1188]: Should be more efficient to just do a `.unstable_sort()` followed by `.uniq()`

File: helix-term/src/commands.rs [hunk 6]
```diff
@@ -1169,6 +1171,27 @@ fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Rege
...
+                .take(20)
```
Comment (archseer) [inline helix-term/src/commands.rs:1189]: Should we limit the amount of items? The rendering UI will already limit it to a maximum

File: helix-term/src/commands.rs (hunk 4 above)
Comment (cossonleo) [reply helix-term/src/commands.rs:1215]: https://github.com/helix-editor/helix/blob/3eb829e2330fed5ad1c095f8bba44f62361b4943/helix-term/src/ui/prompt.rs#L15

prompt::Competion is required `'static`, so Lifetime here is shorter

File: helix-term/src/commands.rs (hunk 5 above)
Comment (cossonleo) [reply helix-term/src/commands.rs:1188]: That will lost history sort.
It doesn't seem to matter. I'll fix it.

File: helix-term/src/commands.rs (hunk 6 above)
Comment (cossonleo) [reply helix-term/src/commands.rs:1189]: It may affect performance When completions computed compete_fn with clone is large

**Review by archseer — APPROVED** [review]

**Pull request merged**